add counter component

Long-HaHoang opened the pull request:
Focus is the refactoring.
Had to move the card to it's own component.
Related to UserStory → #2

RobinWitt commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,88 @@
+  StyledListItem,
+  StyledArticle,
+  StyledDiv,
+  StyledH2,
```
If you style components that are specific to this page, you should try to be more descriptive.
Example: instead of "StyledH2", use "StyledProductTitle" or something like that.
However, if you find this styling should be universal to _all_ your h2's, the styling shouldn't take place in your component

Long-HaHoang replied:
I had to rename the components. It's not semantic enough.
Will modify it accordingly.
But all styling are temporary.

RobinWitt commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,88 @@
+        <StyledDiv>
+          <StyledH2>{product.title}</StyledH2>
+          <StyledParagraph>{product.price}€</StyledParagraph>
+          <WrapperDiv>
```
What is this div doing, other than wrapping your form? Are you sure you need it?

Long-HaHoang replied:
- [x] Removed unused WrapperDiv styled component

RobinWitt reviewed (commented):
I left you two comments, everything else looks good to me!

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+        <span>
+          <h2>{product.title}</h2>
+          <p>{product.price}€</p>
+          <form onSubmit={handleSubmit}>
```
You don't need the form here. It's semantic not correct 'cause it's not really a form.

Long-HaHoang replied:
Yeah you are I removed the input field.

1. It was not part of my userstory
2. Input number caused me some trouble ( Already go some ideas to solve it later, if really need it)

- [x] <form> → <div> (styled will replaced with a grid layout later)
- [x] replaced handleSubmit → handleClick

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+              </button>
+            </span>
+            <span>
+              <button type="submit">
```
This button then is just a normal button with a normal `onClick` method. You can call the `handleSubmit` function by clicking.
Then you don't need the event and `event.preventDefault()` in your `handleSubmit` function and you can name the function something like `addToCard`

Long-HaHoang replied:
- [x] FIX: replaced `handleSubmit` → `addToCart` on a onClick

Long-HaHoang commented:
Removed most the styling, I will change those styling to a grid layout in a later userstory.
For now this is the structure for the temporary style.

- [x] Removed most the styling
- [x] Created components for Icons
- [x] Add counter to header
- [x] Add functionality to counter component
- [x] Add functionality to add-to-cart submit

![Card_Layout_temp](https://user-images.githubusercontent.com/117607177/214573950-ec732a4e-132c-4941-bb8e-245eb188e290.png)

niklasrammelkamp commented on components/ItemCard/index.js:
```diff
@@ -0,0 +1,3 @@
+import { ItemCard } from "./ItemCard";
+
+export default ItemCard;
```
Why do you have this file? 
Either you delete this file or you just paste everything from the ItemCard.js into this file and delete the ItemCard.js

Long-HaHoang replied:
This is called a barrel file.
Usually you bundle multiple files together to export it with one file.
I'm using it to keep my `import` cleaner
Usually I work with index.js files, but if open multiple index.js files it kind looks messy in my VSCode Tab.

In this case I can just import files normally.
`import ItemCard from "@/components/ItemCard"`

Long-HaHoang replied:
Kept the barrel function only for Icons
- [x] Removed barrel function for components

niklasrammelkamp commented on pages/index.js:
```diff
@@ -1,14 +1,18 @@
 
 export default function HomePage() {
+  const [cartTotal, setCartTotal] = useState(0);
+
```
```suggestion
function handleCartTotal(counter) {
      setCartTotal(cartTotal + counter);
    
  }
```

Long-HaHoang replied:
I yes I totally forgot the controlled states.
Thanks a lot 👍

Long-HaHoang replied:
- [x] ADD: handle function
- [x] FIX: props pass-through

niklasrammelkamp commented on pages/index.js:
```diff
@@ -1,14 +1,18 @@
+      <ShopHeader cartTotal={cartTotal} />
       <main>
-        <Gallery />
+        <Gallery setCartTotal={setCartTotal} cartTotal={cartTotal} />
```
Please never pass the set-function to other components. Always keep them in the file where you initialize them. 
I'm not sure about the following but it seems like you pass the cartTotal and you never use it, so you don't need to pass it. 

```suggestion
        <Gallery onSetCartTotal={handleCartTotal}  />
```

Long-HaHoang replied:
- [x] Replaced pass through of setter with handle function

niklasrammelkamp commented on components/Gallery/Gallery.js:
```diff
@@ -1,39 +1,28 @@
+import * as Styled from "@/components/Gallery/Gallery.styles.js";
 
-export function Gallery() {
+export function Gallery({ setCartTotal, cartTotal }) {
```
```suggestion
export function Gallery({ onSetCartTotal, cartTotal }) {
```

Long-HaHoang replied:
- [x] Fixed

niklasrammelkamp commented on components/Gallery/Gallery.js:
```diff
@@ -1,39 +1,28 @@
+            <ItemCard
+              key={product.id}
+              product={product}
+              setCartTotal={setCartTotal}
```
```suggestion
              onSetCartTotal={onSetCartTotal}

Long-HaHoang replied:
- [x] Fixed

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+// Import of SVG Icon Components
+import * as Icon from "@/components/Icons";
+
+export function ItemCard({ product, setCartTotal, cartTotal }) {
```
```suggestion
export function ItemCard({ product, onSetCartTotal, cartTotal }) {
```

Long-HaHoang replied:
- [x] Fix: replaced it with handle function

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+    setCounter(newCounter);
+  }
+
+  function handleSubmit(event) {
+    event.preventDefault();
+    if (counter > 0) {
+      setCartTotal(cartTotal + counter);
+      setCounter(0);
+    }
+  }
```
```suggestion
  function addToCart() {
    if (counter > 0) {
      onSetCartTotal(counter);
      setCounter(0);
    }
  }
```

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+              </button>
+            </span>
+            <span>
+              <button type="submit">
```
```suggestion
              <button type="button" onClick={addToCard}>
```

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+export function ItemCard({ product, setCartTotal, cartTotal }) {
+  const [counter, setCounter] = useState(0);
+
+  function increaseCounter(oldCounter) {
+    const newCounter = oldCounter === 99 ? oldCounter : oldCounter + 1;
+    setCounter(newCounter);
+  }
```
```suggestion
  function increaseCounter() {
    const newCounter = counter === 99 ? counter : counter + 1;
    setCounter(newCounter);
  }
```

Long-HaHoang replied:
A left over of my testing
- [x] Fixed

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+    setCounter(newCounter);
+  }
+
+  function decreaseCounter(oldCounter) {
+    const newCounter = oldCounter === 0 ? oldCounter : oldCounter - 1;
+    setCounter(newCounter);
+  }
```
```suggestion
  function decreaseCounter() {
    const newCounter = counter === 0 ? counter : counter - 1;
    setCounter(newCounter);
  }
```

Long-HaHoang replied:
Same as before
- [x] Fixed

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+          <p>{product.price}€</p>
+          <form onSubmit={handleSubmit}>
+            <span>
+              <button type="button" onClick={() => decreaseCounter(counter)}>
```
```suggestion
              <button type="button" onClick={decreaseCounter}>
```

Long-HaHoang replied:
- [x] Fixed

niklasrammelkamp commented on components/ItemCard/ItemCard.js:
```diff
@@ -0,0 +1,66 @@
+                <Icon.SmallMinus />
+              </button>
+              <p>{counter}</p>
+              <button type="button" onClick={() => increaseCounter(counter)}>
```
```suggestion
              <button type="button" onClick={increaseCounter}>
```

Long-HaHoang replied:
- [x] Fixed

NikRoe reviewed (approved):
Looks good!
